xds/src/main/java/io/grpc/xds/GcpAuthenticationFilter.java

```diff
@@ -0,0 +1,226 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.grpc.xds;
+
+import com.google.auth.oauth2.ComputeEngineCredentials;
+import com.google.auth.oauth2.IdTokenCredentials;
+import com.google.common.primitives.UnsignedLongs;
+import com.google.protobuf.Any;
+import com.google.protobuf.InvalidProtocolBufferException;
+import com.google.protobuf.Message;
+import io.envoyproxy.envoy.extensions.filters.http.gcp_authn.v3.GcpAuthnFilterConfig;
+import io.envoyproxy.envoy.extensions.filters.http.gcp_authn.v3.TokenCacheConfig;
+import io.grpc.CallCredentials;
+import io.grpc.CallOptions;
+import io.grpc.Channel;
+import io.grpc.ClientCall;
+import io.grpc.ClientInterceptor;
+import io.grpc.CompositeCallCredentials;
+import io.grpc.LoadBalancer.PickSubchannelArgs;
+import io.grpc.Metadata;
+import io.grpc.MethodDescriptor;
+import io.grpc.Status;
+import io.grpc.auth.MoreCallCredentials;
+import io.grpc.xds.Filter.ClientInterceptorBuilder;
+import java.util.LinkedHashMap;
+import java.util.Map;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.function.Function;
+import javax.annotation.Nullable;
+
+/**
+ * A {@link Filter} that injects a {@link CallCredentials} to handle
+ * authentication for xDS credentials.
+ */
+final class GcpAuthenticationFilter implements Filter, ClientInterceptorBuilder {
+
+  static final String TYPE_URL =
+      "type.googleapis.com/envoy.extensions.filters.http.gcp_authn.v3.GcpAuthnFilterConfig";
+
+  public GcpAuthenticationFilter() {
+
+  }
+
+  @Override
+  public String[] typeUrls() {
+    return new String[] { TYPE_URL };
+  }
+
+  @Override
+  public ConfigOrError<? extends FilterConfig> parseFilterConfig(Message rawProtoMessage) {
+    GcpAuthnFilterConfig gcpAuthnProto;
+    if (!(rawProtoMessage instanceof Any)) {
+      return ConfigOrError.fromError("Invalid config type: " + rawProtoMessage.getClass());
+    }
+    Any anyMessage = (Any) rawProtoMessage;
+
+    try {
+      gcpAuthnProto = anyMessage.unpack(GcpAuthnFilterConfig.class);
+    } catch (InvalidProtocolBufferException e) {
+      return ConfigOrError.fromError("Invalid proto: " + e);
+    }
+
+    long cacheSize = 10;
+    // Validate cache_config
+    TokenCacheConfig cacheConfig = gcpAuthnProto.getCacheConfig();
+    if (cacheConfig != null) {
+      cacheSize = cacheConfig.getCacheSize().getValue();
+      if (cacheSize == 0) {
+        return ConfigOrError.fromError(
+            "cache_config.cache_size must be in the range (0, INT64_MAX)");
+      } else if (UnsignedLongs.compare(cacheSize, 0L) < 0) {
+        cacheSize = Long.MAX_VALUE;
+      }
+    }
+
+    GcpAuthenticationConfig config = new GcpAuthenticationConfig(cacheSize);
+    return ConfigOrError.fromConfig(config);
+  }
+
+  @Override
+  public ConfigOrError<? extends FilterConfig> parseFilterConfigOverride(Message rawProtoMessage) {
+    return parseFilterConfig(rawProtoMessage);
```

Comment:
Can just return null here?

Reply:
I followed what [FaultFilter](https://github.com/grpc/grpc-java/blob/master/xds/src/main/java/io/grpc/xds/FaultFilter.java#L179) was doing. I'm unsure about its usecase.

Reply:
You can return null and if there is any problem we'd come to know during e2e testing any way.

Reply:
Oh I missed this comment. I'll do this when we revisit Filter for Audience parsing.

Reply:
The gRFC specifies what to do, and it appears the code here is already correct:
>  If the filter's config message appears in a typed_per_filter_config field, it will be validated as part of the normal resource validation, but the configuration will not actually be used.

This method is that validation, and then `buildClientInterceptor()` is when it is ignored.